### PR TITLE
tls: follow dns matching semantics in match subject alt names

### PR DESCRIPTION
**Pull request by ramaraochavali**
For match subject alt names, for exact match cases, follow dns matching semantics which allows wild card processing.
Risk Level: Low
Testing: Added
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
Thank you @ramaraochavali! @LisaLudique do you mind taking a first pass since you found this issue? cc @rebello95 @junr03 as this also fixes an opened Envoy Mobile issue.

@ramaraochavali one small quest: can you maybe do a doc update to make it more clear what the user should configure if they want to do a wildcard match?

**Comment by ramaraochavali** [comment]
@mattklein123 
>  can you maybe do a doc update to make it more clear what the user should configure if they want to do a wildcard match?

Was thinking about it but was n't sure what to update because the new field makes it very clear with string match. The old field doc does not specifically talk about this special DNS matching (which I think people would have understood because there were no questions asked till now AFAIK :-)) . But if you have suggestions to make it better, would be happy to add.

**Comment by mattklein123** [comment]
Maybe just a note here: https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/auth/cert.proto#auth-certificatevalidationcontext in the section on match_subject_alt_names which describes an example of how to do a wildcard match? (Specify exact match with the wildcard match string)

**Comment by junr03** [comment]
Yep definitely relevant to https://github.com/lyft/envoy-mobile/issues/630. I should have opened a parallel issue here. Thanks for fixing @ramaraochavali. I'll review after @LisaLudique takes a pass.

File: api/envoy/api/v2/auth/cert.proto [hunk 1]
```diff
@@ -289,6 +289,16 @@ message CertificateValidationContext {
...
+  // When a certificate has wildcard DNS SAN entries, to match a specific client, it should be
+  // configured with exact match type in the string matcher. For example if the certificate has
+  // "\*.example.com" as DNS service entry, to allow only "api.example.com", it should be configured
+  // as shown below.
+  //
+  // .. code-block:: yaml
+  //
+  // match_subject_alt_names:
+  //   exact: "app.example.com"
+  //
```
Comment (LisaLudique) [inline api/envoy/api/v2/auth/cert.proto]: These documentation changes are great at explaining the necessary configuration for matching with wildcard SAN entries!

A few minor nits:
- link/add ref to StringMatcher/exact_match (e.g. :ref:\`string matcher <envoy_api_msg_type.matcher.StringMatcher>\`)
- I'm guessing "DNS service entry" should be "DNS SAN entry"?
- "app.example.com" should be "api.example.com"

File: source/extensions/transport_sockets/tls/context_impl.cc [hunk 2]
```diff
@@ -690,7 +690,12 @@ bool ContextImpl::matchSubjectAltName(
   for (const GENERAL_NAME* general_name : san_names.get()) {
     const std::string san = generalNameAsString(general_name);
     for (auto& config_san_matcher : subject_alt_name_matchers) {
-      if (config_san_matcher.match(san)) {
+      // For DNS SAN, if it exact, we have to follow DNS matching semantics.
```
Comment (LisaLudique) [inline source/extensions/transport_sockets/tls/context_impl.cc]: nit: replace 'it' with 'the StringMatcher type is exact'

**Pull request merged**